rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc

```diff
@@ -0,0 +1,275 @@
+}
+
+bool isImplicitDefinition(start[Module] _, rel[loc use, loc def] useDef, Define _: <_, _, _, variableId(), defined, _>) = defined in useDef.use; // use of name at implicit declaration loc
+
+bool isImplicitDefinition(start[Module] m, rel[loc, loc] _, Define _: <_, _, _, patternVariableId(), defined, _>) {
+    visit (m) {
+        case qualifiedName(qn):
+            if (locationOfName(qn) == just(defined)) return true;
+        case multiVariable(qn):
+            if (locationOfName(qn) == just(defined)) return true;
+        case variableBecomes(n, _):
+            if (locationOfName(n) == just(defined)) return true;
+    }
+
+    return false;
+}
+
+default bool isImplicitDefinition(start[Module] _, rel[loc, loc] _, Define _) = false;
+
+void checkCausesCapture(TModel tm, start[Module] m, set[Define] currentDefs, set[loc] currentUses, set[Define] newDefs) {
+    set[Define] newNameImplicitDefs = {def | Define def <- newDefs, isImplicitDefinition(m, tm.useDef, def)};
+    set[Define] newNameExplicitDefs = {def | Define def <- newDefs, !isImplicitDefinition(m, tm.useDef, def)};
+
+    // Will this rename turn an implicit declaration of `newName` into a use of a current declaration?
+    rel[loc use, loc oldDef, loc newDef] implicitDeclBecomesUseOfCurrentDecl =
+        {<nD.defined, nD.defined, cD> | Define nD <- newNameImplicitDefs
+                                      , <loc cS, loc cD> <- currentDefs<scope, defined>
+                                      , isContainedIn(nD.defined, cS)
+        };
+
+    // Will this rename turn a use of <oldName> into a use of an existing declaration of <newName> (shadowing)?
+    rel[loc use, loc oldDef, loc newDef] explicitDeclShadowsCurrentDecl =
+        {<cU, cD.defined, nD.defined> | Define nD <- newNameExplicitDefs
+                                      , loc cU <- currentUses
+                                      , Define cD <- getDefines(tm, cU)
+                                      , isContainedIn(cU, nD.scope)
+                                      , isStrictlyContainedIn(nD.scope, cD.scope)
+        };
+
+    throwIfNotEmpty(CausesCapture, implicitDeclBecomesUseOfCurrentDecl + explicitDeclShadowsCurrentDecl);
+}
+
+void checkEditableDefinitions(set[Define] currentDefs) {
+    bool isEditableLocation(loc l) = l.scheme in {"file", "cwd", "home", "memory", "project"};
+    throwIfNotEmpty(DeclarationsNotEditable, {def | loc def <- currentDefs<defined>, !isEditableLocation(def)});
+}
+
+void checkLegalRename(TModel tm, start[Module] m, loc cursorLoc, set[Define] currentDefs, set[loc] currentUses, str newName) {
+    set[Define] newNameDefs = {def | def <- tm.defines, def.id == newName};
+    try {
+        checkEditableDefinitions(currentDefs);
+        checkLegalName(newName);
+        checkCausesDoubleDeclaration(currentDefs, newNameDefs);
+        checkCausesCapture(tm, m, currentDefs, currentUses, newNameDefs);
+    } catch IllegalRenameReason r: {
```

Comment:
I'm not sure I truly like this style of using exceptions for "regular" control flow. I get the beautify, but there are some constrains to this style:

1. it only allows for a single message
2. we have to be careful about other exceptions.

if every checker would return a set of `IllegalRenameReason` we can plus all the calls together, and if the result is non-empty, throw it is an illegal rename. It might also remove the need for `throwIfNotEmpty` as we can use the comprehension directly (with a small change of the signature of the `IllegalRenameReason` ADT I guess).